### PR TITLE
Implementation of ResNeXt

### PR DESCRIPTION
**Pull request by theabhirath**
This code implements [ResNeXt](https://arxiv.org/pdf/1611.05431.pdf), a convolutional neural network that is often used for image classification and as backbone for image segmentation tasks.

### CODE REVIEW

**Comment by theabhirath** [comment]
The `show` function doesn't seem to be working as expected for large models on this repo, however.

Screenshot 2022-01-15 at 8 13 37 PM

**Comment by theabhirath** [comment]
Resolved previous issues, this now seems good to go.

**Comment by darsnack** [comment]
I will need to look through the paper later tonight to make sure everything is good before merging, but on a first pass, this looks great.

The only changes for now are to fix minor formatting issues like code lines not being lined up. Also, can you update the constructors for ResNeXt50 etc. to match #99?

File: src/resnext.jl
Comment (DhairyaLGandhi) [inline src/resnext.jl:40]:
```suggestion
function resnext(connection = (x, y) -> @. relu(x) + relu(y);
                             channel_config, block_config, cardinality, nclasses = 1000)
```

File: src/resnext.jl
Comment (DhairyaLGandhi) [inline src/resnext.jl:117]: See #99 for the simpler API

File: src/resnext.jl
Comment (DhairyaLGandhi) [inline src/resnext.jl:88]: Seems to me that the API is backwards here. `channel_config`, `block_config` and `cardinality` need to be positional, and have convenience constructors for skipping the `connection`.

**Review by DhairyaLGandhi — COMMENTED** [review]
This is looking great so far! Could you also add this to the tests please? Looking forward to the transformers!

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:24]: I believe the ResNeXt block is best understood as # of in channels, cardinality, and bottleneck width (see Fig. 3c in the paper). As such, the API for the block should probably be more like the following:
```julia
function resnextblock(inplanes, outplanes, cardinality, width, downsample = false)
    stride = downsample ? 2 : 1
    inner_channels = cardinality * width

    return Chain(conv_bn((1, 1), inplanes, inner_channels; stride = 1, bias = false)...,
                 conv_bn((3, 3), inner_channels, inner_channels;
                         stride = stride, pad = 1, bias = false, groups = cardinality)...,
                 conv_bn((1, 1), inner_channels, outplanes; stride = 1, bias = false)...)
end
```

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:66]: As a result of my comment about `resnextblock`, we don't really need `channel_config` here. This is one of the features of ResNeXt...it simplifies the design quite a bit by thinking more generically. In the paper, `channel_config == [1, 1, 2]` always, so we could just hardcode the fact that the output channels doubles at the end of the block. But I think future-proof this a bit by adding a `channel_multiplier`. Here's how I would do it.

Also, note a bug that `inplanes = 64` to start.

```suggestion
function resnext(cardinality, width, channel_multiplier = 2, connection = (x, y) -> @. relu(x) + relu(y);
                 block_config, nclasses = 1000)
    inplanes = 64
    baseplanes = 128
    layers = []
    append!(layers, conv_bn((7, 7), 3, inplanes; stride = 2, pad = (3, 3)))
    push!(layers, MaxPool((3, 3), stride = (2, 2), pad = (1, 1)))
    for (i, nrepeats) in enumerate(block_config)
        # output planes within a block
        outplanes = baseplanes * channel_multiplier
        # push first skip connection on using first residual
        # downsample the residual path if this is the first repetition of a block
        push!(layers, Parallel(connection, resnextblock(inplanes, outplanes, cardinality, width, i != 1),
                                           skip_projection(inplanes, outplanes, i != 1)))
        # push remaining skip connections on using second residual
        inplanes = outplanes
        for _ in 2:nrepeats
            push!(layers, Parallel(connection, resnextblock(inplanes, outplanes, cardinality, width, false),
                                               skip_identity(inplanes, outplanes, false)))
        end
        baseplanes = outplanes
    end

    return Chain(Chain(layers...),
                 Chain(AdaptiveMeanPool((1, 1)), flatten, Dense(inplanes, nclasses)))
end
```

File: src/resnext.jl
Comment (darsnack) [reply src/resnext.jl:88]: That's because this mirrors the `resnet` API (as it should for consistency) and the `resnet` API contains additional positional arguments that are more important, but those arguments are dropped here. In any case, I think the API needs some more significant changes since ResNeXt is specified in terms of cardinality and width, not a channel configuration. I've suggested an alternate API where the relevant positional args are given first.

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:84]: Accordingly, this should be `ResNeXt(block_config, cardinality, width; nclasses = 1000)`.

**Review by darsnack — CHANGES_REQUESTED** [review]
Despite the big diffs here, I think the initial implementation is solid. The major changes are to the API which should be based primarily on cardinality and bottleneck width, since that is how various ResNeXt configurations are formed.

Also, +1 on @DhairyaLGandhi's suggestions for simplifying the constructors and adding tests.

Thanks for this PR, it will make a great addition to the package!

File: src/resnext.jl
Comment (DhairyaLGandhi) [reply src/resnext.jl:66]: I'd rather avoid hard coding it if there are variants of interest that diverge from the default. It keeps the design simpler by hoisting the `channel_multiplier` outside the package. If the dimensions don't align, that would error on the forward pass, so we should be fine to report those.

Comment (darsnack) [reply src/resnext.jl:66]: Not sure what you mean exactly, but my suggestion is not to hardcode it.

File: src/resnext.jl
Comment (theabhirath) [reply src/resnext.jl:24]: Yeah, I thought of doing away with the original ResNet style API but this is my first contribution so was a little unsure of how much I could change. The cardinality x width specification is much more characteristic for ResNeXt, I agree.

**Comment by theabhirath** [comment]
Okay so while writing the tests I noticed they're still in the style of the old API for the ResNet, so should I take the liberty of updating it so that the ResNet and ResNeXt tests match the new API as proposed in #99 ?

**Comment by theabhirath** [comment]
There is also the addition problem that since the struct and the function to integrate with it as well as the user API have the same name, VSCode IntelliSense seems to be messing up the display of documentation on hover - it's showing both of them back-to-back which is not ideal. Simply changing the function name to have a small n (the struct and the function to integrate with it) fixes this, so if that's okay I can make that change as well.

**Comment by darsnack** [comment]
> Okay so while writing the tests I noticed they're still in the style of the old API for the ResNet, so should I take the liberty of updating it so that the ResNet and ResNeXt tests match the new API as proposed in #99 ?

Yes, just for ResNeXt. The other models should stay as is, since they will be handled in #99.

> VSCode IntelliSense seems to be messing up the display of documentation on hover - it's showing both of them back-to-back

Are you referring to the docstring for `ResNeXt` with the new API? It will show both the docstring above the struct as well as above the convenience constructor for pre-trained models? This is not a problem. In Julia, it is typical for a single function (i.e. `ResNeXt`) to have multiple methods (i.e. the default constructor and the convenience constructor). When a user types `ResNeXt` in their code, they may want either constructor which is why the VSCode extension displays both. Sometimes, we merge all of these definitions into a single docstring above the struct, but if they are different enough (as is the case here), then we would prefer separate docstrings. I would not change one to have a lower case "n" and the other capitalized (maybe I misunderstood the suggestions, so feel free to clarify).

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:24]: I don't believe the `widen_factor` should be considered here. The inner channel dimension is always `cardinality * width`, and it is independent of the input and output channels. See Fig. 3c in the paper: the input and output channels are 256, and the inner channels is 32*4 = 128.

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:41]: Make sure to follow the style of the rest of the code base. We don't indent by only 1 on a new line. `block_config` should line up with `cardinality` here.

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:111]: `config` should be a positional argument, and `cardinality` and `width` should be part of the configuration dictionary.

File: test/runtests.jl
Comment (darsnack) [inline test/runtests.jl:54]: Try not to change the other code.

File: test/runtests.jl
Comment (darsnack) [inline test/runtests.jl:63]:
```suggestion
    m = ResNeXt(config = depth)
```

**Review by darsnack — CHANGES_REQUESTED** [review]

**Comment by darsnack** [comment]
Looking good, just a few more things to clear up.

File: src/resnext.jl
Comment (theabhirath) [reply src/resnext.jl:24]: That doesn't really match the image though. In that case, the inner channels should always be 128, whereas the image explicitly shows the inner channels increasing for every block. And plus that configuration does not match the parameters specified in the paper, only this one does. The params and FLOPs line up with this specific implementation. But let me know if I'm missing something

File: test/runtests.jl
Comment (theabhirath) [reply test/runtests.jl:54]: Whoops, must've snuck in there accidentally, apologies

Comment (darsnack) [reply test/runtests.jl:54]: No problem!

File: test/runtests.jl
Comment (theabhirath) [reply test/runtests.jl:63]: I kept the original style because both `model` and `m` are being used for tests...should I just do away with the model definition and replace it everywhere by `ResNeXt`?

File: src/resnext.jl
Comment (theabhirath) [reply src/resnext.jl:111]: This one is tricky, because what the original ResNeXt paper allows for is changes to cardinality and width even while maintaining a single configuration of layers - they've given versions that are 64x4d, for instance. I wanted to allow for that. I'll change config to be a positional argument, though.

File: src/resnext.jl
Comment (darsnack) [reply src/resnext.jl:24]: Oh you're right, I was completely focused on the figures, but the table does indeed have an increasing inner dimension. What I would suggest then is still keeping the `inner = cardinality * width` approach since that matches how the bottleneck is described in the paper. Then in main `resnext` function, we can widen the `width` passed in by `widen_factor` just like we do for `baseplanes`.

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:60]:
```suggestion
        baseplanes = outplanes
        width *= widen_factor
```

**Review by darsnack — CHANGES_REQUESTED** [review]

File: test/runtests.jl
Comment (darsnack) [reply test/runtests.jl:63]: Yeah just for ResNeXt tests. The old style was only necessary cause the function names were different for each configuration. In the new approach, we know it will always be `ResNeXt` with the depth argument varying.

**Comment by theabhirath** [comment]
Should I also shift the ResNet utility functions `skip_identity` and `skip_projection` over to `utilities.jl`? I've used them in ResNeXt and a couple of other models that I'm drafting without it being clear that they're actually in the ResNet file

**Comment by darsnack** [comment]
Yes, that makes sense

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:22]:
```suggestion
    return Chain(conv_bn((1, 1), inplanes, hidden_channels; stride = 1, bias = false)...,
                 conv_bn((3, 3), hidden_channels, hidden_channels;
                         stride = stride, pad = 1, bias = false, groups = cardinality)...,
                 conv_bn((1, 1), hidden_channels, outplanes; stride = 1, bias = false)...)
```

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:2]:
```suggestion
    resnextblock(inplanes, outplanes, cardinality, width, downsample = false)
```

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:40]:
```suggestion
                 block_config, nclasses = 1000)
```

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:27]:
```suggestion
    resnext(cardinality, width, widen_factor, connection = (x, y) -> @. relu(x) + relu(y);
            block_config, nclasses = 1000)
```

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:34]: Add `widen_factor`

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:52]:
```suggestion
        push!(layers, Parallel(connection, resnextblock(inplanes, outplanes, cardinality, width, i != 1),
                                           skip_projection(inplanes, outplanes, i != 1)))
```

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:57]: Same as above

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:64]:
```suggestion
    return Chain(Chain(layers...),
                 Chain(AdaptiveMeanPool((1, 1)), flatten, Dense(inplanes, nclasses)))
```

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:75]: Add `block_config`

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:103]:
```suggestion
Create a ResNeXt model with specified configuration. Currently supported values for `config` are (50, 101).
```

File: test/runtests.jl
Comment (darsnack) [inline test/runtests.jl:54]:
```suggestion
                         channel_config = [1, 1],
                         block_config = [2, 2, 2, 2])
```

**Review by darsnack — CHANGES_REQUESTED** [review]
Okay looks great, mostly minor formatting changes at this point and fixing docstrings. One annoying change is to use an indent of 2 spaces like the rest of the repo.

**Comment by theabhirath** [comment]
I _think_ that's everything....

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:2]:
```suggestion
    resnextblock(inplanes, outplanes, cardinality, width, downsample = false)
```

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:26]:
```suggestion
    resnext(cardinality, width, widen_factor = 2, connection = (x, y) -> @. relu(x) + relu(y);
```

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:70]:
```suggestion
    ResNeXt(cardinality, width; block_config, nclasses = 1000)
```

File: src/resnext.jl
Comment (darsnack) [inline src/resnext.jl:104]:
```suggestion
    ResNeXt(config::Int = 50; cardinality = 32, width = 4, pretrain = false, nclasses = 1000)
```

**Review by darsnack — CHANGES_REQUESTED** [review]
Almost... 😅

**Review by darsnack — APPROVED** [review]
Looks great! Will merge once CI is green. Thank you @theabhirath for seeing this through! Looking forward to more contributions.

**Comment by theabhirath** [comment]
> Looks great! Will merge once CI is green. Thank you @theabhirath for seeing this through! Looking forward to more contributions.

Thanks for walking me through! I'm gonna be trying for some other missing models like MobileNet and EfficientNet next, I'll need the help 😅.

> Almost... 😅. FluxML's style is needlessly frustrating to follow.

By any chance, is there any auto-formatter? I already changed VSCode to have indent as two spaces for the future, but for the other style guidelines?

**Comment by darsnack** [comment]
> Thanks for walking me through! I'm gonna be trying for some other missing models like MobileNet and EfficientNet next, I'll need the help 😅.

Looking forward to it!

> By any chance, is there any auto-formatter? I already changed VSCode to have indent as two spaces for the future, but for the other style guidelines?

Unfortunately, no. But the only other main one I noticed was when you break a long list of arguments onto a new line. The next line should line up with the first argument in the previous line (not just a 1 indent like other languages).

**Comment by DhairyaLGandhi** [comment]
Great job @theabhirath ! Also see https://github.com/pxl-th/EfficientNet.jl and https://github.com/pxl-th/Segmentation.jl/issues/1 for EfficientNet. Would love to see these in Metalhead!

**Pull request merged**
